### PR TITLE
Ee 766

### PR DESCRIPTION
**Pull request by vinnicombe**

### CODE REVIEW

File: app/views/AmountsAlreadyInCodeView.scala.html [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+                    <tr>
+                        <th>@messages("amountsAlreadyInCode.tableHeading1")</th>
+                        <th>@messages("amountsAlreadyInCode.tableHeading2")</th>
+                    </tr>
+                </thead>
+                <tbody>
+                    @taxYearSelection.map(
+                        taxYear =>
+                            <tr>
+                                <td id={taxYear.toString}>
```
Comment (MatthewAlexanderBeck) [inline app/views/AmountsAlreadyInCodeView.scala.html]: Does this need to be toString?

File: app/views/AmountsAlreadyInCodeView.scala.html [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+                                    {taxYearText(getTaxYear(taxYear))}
+                                </td>
+                                <td id={taxYear.toString + "-amount"} class="row">
```
Comment (MatthewAlexanderBeck) [inline app/views/AmountsAlreadyInCodeView.scala.html]: Same here?

File: test/views/CannotClaimEmployerContributionViewSpec.scala [hunk 3]
```diff
@@ -17,7 +17,6 @@
 package views
```
Comment (garethgooch) [inline test/views/CannotClaimEmployerContributionViewSpec.scala]: Can we add an application stop on this

File: test/views/SummarySubscriptionsViewSpec.scala [hunk 4]
```diff
@@ -18,6 +18,7 @@ package views
```
Comment (garethgooch) [inline test/views/SummarySubscriptionsViewSpec.scala]: application stop

File: test/views/UpdateYourAddressViewSpec.scala [hunk 5]
```diff
@@ -17,8 +17,7 @@
 package views
```
Comment (garethgooch) [inline test/views/UpdateYourAddressViewSpec.scala]: application stop

File: test/views/UpdateYourEmployerInformationViewSpec.scala [hunk 6]
```diff
@@ -17,7 +17,6 @@
 package views
```
Comment (garethgooch) [inline test/views/UpdateYourEmployerInformationViewSpec.scala]: application stop

File: test/views/CannotClaimEmployerContributionViewSpec.scala (hunk 3 above)
Comment (vinnicombe) [reply test/views/CannotClaimEmployerContributionViewSpec.scala]: None of the rest of the view specs have application.stop()

File: test/views/SummarySubscriptionsViewSpec.scala (hunk 4 above)
Comment (vinnicombe) [reply test/views/SummarySubscriptionsViewSpec.scala]: None of the rest of the view specs have application.stop()

File: test/views/UpdateYourAddressViewSpec.scala (hunk 5 above)
Comment (vinnicombe) [reply test/views/UpdateYourAddressViewSpec.scala]: None of the rest of the view specs have application.stop()

File: test/views/UpdateYourEmployerInformationViewSpec.scala (hunk 6 above)
Comment (vinnicombe) [reply test/views/UpdateYourEmployerInformationViewSpec.scala]: None of the rest of the view specs have application.stop()

File: app/controllers/CheckYourAnswersController.scala [hunk 7]
```diff
@@ -48,31 +49,38 @@ class CheckYourAnswersController @Inject()(
...
+                      rows = Seq(
+                        cyaHelper.whichSubscription(getTaxYear(taxYear).toString, index, psub),
+                        cyaHelper.subscriptionAmount(getTaxYear(taxYear).toString, index, psub),
+                        cyaHelper.employerContribution(getTaxYear(taxYear).toString, index, psub),
+                        cyaHelper.expensesEmployerPaid(getTaxYear(taxYear).toString, index, psub)
+                      ).flatten,
+                      messageArgs = Seq(getTaxYear(taxYear).toString, (getTaxYear(taxYear) + 1).toString): _*
+                    )
+                }
+              )
```
[the remark this answers is not in the document]
Comment (vinnicombe) [reply app/controllers/CheckYourAnswersController.scala]: Up to you. I don't think it's offensively long at this stage.

File: app/forms/AmountsAlreadyInCodeFormProvider.scala [hunk 8]
```diff
@@ -17,14 +17,21 @@
...
+import models.NpsDataFormats._
+
+class AmountsAlreadyInCodeFormProvider @Inject() extends Mappings {
+
+  def apply(userAnswers: UserAnswers): Form[Boolean] = {
 
-class IsYourDataCorrectFormProvider @Inject() extends Mappings {
+    val errorMessageKey: String = userAnswers.get(NpsData) match {
+      case Some(npsData) => if (npsData.size == 1) "single" else "multiple"
+      case _ => ""
```
[the remark this answers is not in the document]
Comment (vinnicombe) [reply app/forms/AmountsAlreadyInCodeFormProvider.scala]: If NpsData is not present, the view will never load and the form provider will not be called.

File: app/models/NpsDataFormats.scala [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+      }
+
+      def reads(json: JsValue): JsResult[Map[Int, Seq[EmploymentExpense]]] = {
+        json.validate[Map[String, Seq[EmploymentExpense]]].map(_.map {
+          case (key, value) => key.toInt -> value
+        })
+      }
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (vinnicombe) [reply app/models/NpsDataFormats.scala]: Will add

File: app/utils/CheckYourAnswersHelper.scala [hunk 10]
```diff
@@ -18,32 +18,58 @@ package utils
...
 class CheckYourAnswersHelper(userAnswers: UserAnswers)(implicit messages: Messages) {
 
-  def tellUsWhatIsWrong: Option[AnswerRow] = userAnswers.get(TellUsWhatIsWrongPage) map {
-    x =>
+  def taxYearText(taxYear: TaxYearSelection): String =
+    messages(s"taxYearSelection.$taxYear", getTaxYear(taxYear).toString, (getTaxYear(taxYear) + 1).toString)
+
+  def npsDataFormatted(npsData: Map[Int, Seq[EmploymentExpense]]): String = {
+
+    val sortedData: Seq[(Int, Seq[EmploymentExpense])] = ListMap(npsData.toSeq.sortWith(_._1 > _._1): _*).toSeq
```
[the remark this answers is not in the document]
Comment (vinnicombe) [reply app/utils/CheckYourAnswersHelper.scala]: Moved

**Pull request merged**